datastructs/ring.c

```diff
@@ -0,0 +1,76 @@
+// ring.c : 
+/* A ring buffer for multi-threaded programs. A single producer pushes
+* objects into the ring, while a single consumer pops them off for
+* processing. If the ring is full (i.e., the consumer is slower than the
+* producer), then the push operation fails.
+*/
+
+
+
+
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "ring.h"
+
+
+
+
+/* Create a ring buffer with the specified size. Return the ring or NULL
+if there is not enough memory to create. */
+struct ring* ring_create(int size)
+{
+	if (size <= 0)  return NULL;
+
+
+
+	struct ring *rb = (struct ring*) malloc( sizeof(struct ring) );
+	rb->push = 0;
+	rb->pop = 0;
+
+	rb->size = size;
+	rb->count = 0;
+
+	rb->buffer = (void*) malloc( size*sizeof(int) );
+
+	return rb;
+};
+
+/* Add an entry to the ring.
+Return 0 on success, or a sensible error code if ring is full
+*/
+int ring_push(struct ring *rb, void* data)
+{ 
+	if(rb->count >= rb->size)
+		return -1;
+	else
+	{
+		*( (int*)( rb->buffer ) + rb->push ) = *(int*) data;
```

Comment:
My preference is to avoid pointers unless necessary.  You should be able to use array access style to your `rb->buffer`.  Try to rewrite this with something like: `rb->buffer[rb->push] = data`.